Remove unused optional dep to avoid circular dep

Awkanimus opened the pull request:
Removing `ethereal` from the `optional_depends` because it is not used in moretrees and it creates a circular dependency.

The dependency cycle this solves is currently ethereal --> bakedclay --> flowerpot --> moretrees --> ethereal.

Niklp09 commented:
https://github.com/mt-mods/moretrees/pull/11#discussion_r1002973356

Was this fixed by TenPlus1?

wsor4035 commented:
> [#11 (comment)](https://github.com/mt-mods/moretrees/pull/11#discussion_r1002973356)
> 
> Was this fixed by TenPlus1?

nope

wsor4035 commented:
to clarify the label, tenplus1 is to blame here for still nuking things unless mods dep on ethereal to load after it

Niklp09 commented:
relevant yl issue https://gitea.your-land.de/your-land/bugtracker/issues/3013

wsor4035 commented:
~~probably~~ should make an issue from tracking the status so we dont forget
^ ill do this

wsor4035 commented:
see https://github.com/mt-mods/moretrees/issues/22

Awkanimus commented:
Forgot to comment. Thanks for the clarification. I'll watch the related bug.

wsor4035 reviewed (changes requested):
see https://github.com/mt-mods/moretrees/issues/22#issuecomment-1345134188 - mine and buckaroos comments

Awkanimus commented:
Updated as requested. Notes:
1. Made generic so the same mod_dependency_checks.lua could be used in other projects
2. Dependencies can be marked as "fatal" in which case server does not start when check fails
3. Dependencies can be marked as "optional" which skips the check if the dependency is missing
4. logs at the "error" level are visible to players, all others are not. A summary is displayed to the player (who may be less technical) and an explanation is dropped in debug.txt.
5. I did try to do this with sending chat messages instead of logs and it didn't work so well; I think the players dont exist while the server is loading or something (which kinda makes sense). Given the desire to halt a server from starting up, I moved on.

Awkanimus commented:
Testing notes:
1. Tested world with and without ethereal loaded and ensured world loaded.
2. Tested in world with recent ethereal loaded and ensured world loaded.
3. Changed the required version by adding "99" to the front and validated that error message was provided
4. Changed optional to false and revalidated point 3
5. Changed optional to false and loaded world without ethereal and ensured error (even though this case should have used mod.conf)

Awkanimus commented:
For clarity, coding and testing is all done. Not sure if requesting a rereview is the way forward here but that's what I did

wsor4035 commented:
im sorry, but these latest commits are absolutely ridiculous. nothing more was needed besides  a minimally modified version of the code listed in https://github.com/mt-mods/moretrees/issues/22#issuecomment-1345102761 - additionally, it seems like luacheck wasnt even used here (see failed ci)

Awkanimus commented:
I just noticed you folks use tabs, not spaces. Just a moment.

Awkanimus commented:
Ok @wsor4035, a version more similar to the one in the other conversation is force-pushed. Tabs, not spaces are used as thats what the code around it mostly uses. Luacheck (assuming luacheck's config plus the .luacheckrc settings are enough) passes.

wsor4035 reviewed (approved):
couple of minor things, feel free to reject them or change them if you wish. overall this looks good.

use `minetest.global_exists(name)` or `rawget` rather than just `if ethereal then` since minetest has a builtin version of luacheck which is terrible (actual luacheck doesnt complain rightfully so) because it will think your trying to access an undeclared global (when ethereal isnt installed).
doc: https://github.com/minetest/minetest/blob/981d79157a64468d19f140f5c55d8ccd3855fa71/doc/lua_api.txt#L6573-L6574

the comment `-- Validate that if ethereal exists, that it's version is greater than 20220424` could be changed to mention something like `-- Validate that if ethereal exists, that it's version is greater than 20220424 so it doesnt break this mod` giving a why for the check

wsor4035 commented:
oh, just noticed `moretress` should be `moretrees` in the `error` function

Awkanimus commented:
Ok. The comment is more descriptive, now uses `minetest.global_exists("ethereal")` and s/moretress/moretrees/g.

Awkanimus commented:
Also, manual tests with and without ethereal and with a bad version number rerun and continue to pass. Luacheck passes.

wsor4035 commented:
looks good. approval still stands.

wsor4035 commented:
thank you for your contribution and testing @Awkanimus

Awkanimus commented:
Happy to help. Thanks for the timely responses and constructive feedback.
